tests/test_location_pi.py

```diff
@@ -0,0 +1,508 @@
+        req = {
+            'location_id': self.test_location_01.id,
+            'pi_count_moves': [
+                {
+                    'quants_ids': [1, 2, 3],
+                    'location_id': self.test_location_01.id,
+                    'location_dest_id': self.unknown_location_id
+                }
+            ]
+        }
+
+        with self.assertRaisesRegex(ValidationError,
+                                    'The request has an unknown location'):
+            self.test_location_01._validate_perpetual_inventory_request(req)
+
+    def test03_validate_pi_request_failure_no_adj(self):
+        """ Errors if the request contains only preceeding adjustments """
+        req = {
+            'location_id': self.test_location_01.id,
+            'pi_count_moves': [
+                {
+                    'quants_ids': [1, 2, 3],
+                    'location_id': self.test_location_01.id,
+                    'location_dest_id': self.test_location_02.id
+                }
+            ],
+            'preceding_inventory_adjustments': {
+                'location_id': self.test_location_01.id,
+                'inventory_adjustments': [
+                    {'product_id': 7, 'package_name': 'foo', 'quantity': 0}
+                ]
+            }
+        }
+
+        with self.assertRaisesRegex(ValidationError,
+                                    'You must specify inventory adjustments'):
+            self.test_location_01._validate_perpetual_inventory_request(req)
+
+    def test04_validate_pi_request_failure_unknown_adjustment_location(self):
+        """ Errors if the request contains unknown locations """
+        req = {
+            'location_id': self.test_location_01.id,
+            'inventory_adjustments': [
+                {
+                    'product_id': self.apple.id,
+                    'package_name': self.package_one.name,
+                    'quantity': 4
+                }
+            ],
+            'preceding_inventory_adjustments': {
+                'location_id': self.unknown_location_id,
+                'inventory_adjustments': [
+                    {
+                        'product_id': self.banana.id,
+                        'package_name': self.package_two,
+                        'quantity': 0
+                    }
+                ]
+            }
+        }
+
+        with self.assertRaisesRegex(ValidationError,
+                                    'The request has an unknown location'):
+            self.test_location_01._validate_perpetual_inventory_request(req)
+
+    def test05_process_pi_datetime_no_changes(self):
+        """
+        Both last check & last correct PI datetime are updated if
+        the PIOutcome is empty
+        """
+        outcome = MockPIOutcome(False)
+        old_dt_checked = self.test_location_01.u_date_last_checked
+        old_dt_success = self.test_location_01.u_date_last_checked_correct
+        time.sleep(1)
```

Comment:
do we need this sleep?

Reply:
it ensures it will never fail for state

Reply:
which state? I think by default the dates are empty